### PR TITLE
Addition of PEL message method for dump delete/offload

### PR DESCRIPTION
**Pull request by Swarnendu-R-C**
Addition of PEL message method for dump delete/offload
 
Adding a createPEL method which will log PEL messages.
This method has PEL severity as Informational by default
unless mentioned otherwise

Change-Id: b4e1622407fa9aa372fe0f20f73509cb376f90c5
Signed-off-by: Swarnendu Roy Chowdhury <swarnendu.roy.chowdhury@ibm.com>

### CODE REVIEW

File: bmcstored_dump_entry.cpp [hunk 1]
```diff
@@ -86,7 +99,7 @@ void Entry::initiateOffload(std::string uri)
 
     if (pid > 0)
     {
-        Child::Callback callback = [this](Child&, const siginfo_t* si) {
+        Child::Callback callback = [this, uri](Child&, const siginfo_t* si) {
```
Comment (devenrao) [inline bmcstored_dump_entry.cpp:102]: what is this uri change for, if it is not for this commit, please make a different patch

File: bmcstored_dump_entry.cpp [hunk 2]
```diff
@@ -31,7 +31,9 @@ void Entry::delete_()
         elog<sdbusplus::xyz::openbmc_project::Common::Error::Unavailable>();
     }
 
-    // Delete Dump file from Permanent location
+    // Delete Dump file from Permanent location but before that copy the dump
+    // file name to be put in the PEL message
+    const auto strDumpFileName = path();
```
Comment (devenrao) [inline bmcstored_dump_entry.cpp:36]: what is this change for?

File: dump_utils.cpp [hunk 3]
```diff
@@ -147,5 +147,79 @@ bool isHostQuiesced()
...
+            "xyz.openbmc_project.Logging.Create";
+        constexpr auto loggerService = "xyz.openbmc_project.Logging";
+        std::string pelSeverity =
+            "xyz.openbmc_project.Logging.Entry.Level.Informational";
+        auto itr = sevMap.find(sev);
+        if (itr != sevMap.end())
+            pelSeverity = itr->second;
+
+        sd_bus* pSD_Bus = nullptr;
+        ;
```
Comment (devenrao) [inline dump_utils.cpp]: remove

File: dump_utils.hpp [hunk 4]
```diff
@@ -166,5 +166,34 @@ T readDBusProperty(sdbusplus::bus::bus& bus, const std::string& service,
...
+};
+
+/**
+ * @brief Create a new PEL message for dump Delete/Offload
+ *
+ * @param[in] additionalData - dump id, name and type.
+ * @param[in] sev - severity.
+ * @param[in] errIntf - D-Bus interface name.
+ * @return Returns void on success throws exception on error
+ **/
```
Comment (devenrao) [inline dump_utils.hpp:193]: void createPEL(
    sdbusplus::bus_t& bus,
    const std::unordered_map<std::string, std::string>& additionalData,
    const PelSeverity& sev, const std::string& errIntf);


caller has bus, you can pass that to this method

File: dump_utils.cpp [hunk 5]
```diff
@@ -147,5 +147,79 @@ bool isHostQuiesced()
...
+        sd_bus_default(&pSD_Bus);
```
Comment (devenrao) [inline dump_utils.cpp:192]: get it from caller

File: dump_utils.cpp [hunk 6]
```diff
@@ -147,5 +147,79 @@ bool isHostQuiesced()
...
+    {PelSeverity::ERROR, "xyz.openbmc_project.Logging.Entry.Level.Error"},
+    {PelSeverity::ALERT, "xyz.openbmc_project.Logging.Entry.Level.Alert"}};
+
+void createPEL(
+    const std::unordered_map<std::string, std::string>& additionalData,
+    const PelSeverity& sev, const std::string& errIntf)
+{
+    try
+    {
+        auto bus = sdbusplus::bus::new_default();
```
Comment (devenrao) [inline dump_utils.cpp:180]: unused, you can get this from caller

File: dump_utils.cpp [hunk 7]
```diff
@@ -147,5 +147,79 @@ bool isHostQuiesced()
...
+            throw std::runtime_error("Dump delete/offload PEL not created due "
+                                     "to insufficient info passed");
+        }
+        auto itrToAdditionalData = additionalData.begin();
+        auto retVal = sd_bus_call_method_async(
+            pSD_Bus, nullptr, loggerService, loggerObjectPath,
+            loggerCreateInterface, "Create", nullptr, nullptr, "ssa{ss}",
+            errIntf.c_str(), pelSeverity.c_str(), 3,
+            itrToAdditionalData->first.c_str(),
+            itrToAdditionalData->second.c_str(),
```
Comment (devenrao) [inline dump_utils.cpp:202]: did you ran clang-format on this code?

**Review by devenrao — COMMENTED** [review]
please fix the comments

File: dump_utils.cpp [hunk 8]
```diff
@@ -167,6 +167,16 @@ void createPEL(
 {
     try
     {
+        if (additionalData.empty() || additionalData.size() < 3)
```
Comment (devenrao) [inline dump_utils.cpp:170]: why this change came into this commit shouldn't;t this be in the first commit

File: dump_utils.cpp [hunk 9]
```diff
@@ -167,6 +167,16 @@ void createPEL(
...
+        {
+            log<level::WARNING>(
+                fmt::format("User data map's size(({})) is not sufficient to "
+                            "create a PEL message for dump delete/offload",
+                            additionalData.size())
+                    .c_str());
+            throw std::runtime_error("Dump delete/offload PEL not created due "
+                                     "to insufficient info passed");
+        }
```
Comment (devenrao) [inline dump_utils.cpp:179]: why this change in this commit shoudn;t this be in first commit

**Review by devenrao — COMMENTED** [review]
Commit meesage is wrong, here you are facilitating an utility method to create a pel. Provide appropriate commit message

File: dump_utils.cpp (hunk 6 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:180]: Removed

File: dump_utils.cpp (hunk 3 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp]: `void Entry::delete_()
{
    auto srcDumpID = sourceDumpId();
    auto dumpId = id;

    if ((!offloadUri().empty()) && (phosphor::dump::isHostRunning()))
    {
        log<level::ERR>(fmt::format("Dump offload in progress, cannot delete "
                                    "dump, id({}) srcdumpid({})",
                                    dumpId, srcDumpID)
                            .c_str());
        elog<sdbusplus::xyz::openbmc_project::Common::Error::Unavailable>();
    }

    auto path = std::filesystem::path(RESOURCE_DUMP_SERIAL_PATH) /
                std::to_string(dumpId);

    log<level::INFO>(
        fmt::format(
            "Resource dump delete id({}) srcdumpid({}) and serial path({})",
            dumpId, srcDumpID, path.string().c_str())
            .c_str());

    // Remove resource dump when host is up by using source dump id
    // which is present in resource dump entry dbus object as a property.
    if ((phosphor::dump::isHostRunning()) && (srcDumpID != INVALID_SOURCE_ID))
    {
        try
        {
            phosphor::dump::host::requestDelete(srcDumpID,
                                                TRANSPORT_DUMP_TYPE_IDENTIFIER);
        }
        catch (const std::exception& e)
        {
            log<level::ERR>(fmt::format("Error deleteing dump from host id({}) "
                                        "host id({}) error({})",
                                        dumpId, srcDumpID, e.what())
                                .c_str());
            elog<sdbusplus::xyz::openbmc_project::Common::Error::Unavailable>();
        }
    }

    // Remove Dump entry D-bus object
    phosphor::dump::Entry::delete_();
    log<level::INFO>(
        fmt::format("Resource dump entry with id({}) is deleted", dumpId)
            .c_str());
    try
    {
        std::filesystem::remove_all(path);
    }
    catch (const std::filesystem::filesystem_error& e)
    {
        // Log Error message and continue
        log<level::ERR>(
            fmt::format("Failed to delete dump file({}), errormsg({})",
                        path.string().c_str(), e.what())
                .c_str());
    }
    // Log PEL for dump /offload
    log<level::INFO>("Log PEL for dump delete or offload");

    std::unordered_map<std::string, std::string> additionalData = {
        {"Dump ID", std::to_string(dumpId)},
        {"Filename", path.string()},
        {"Dump Type", "Resource Dump"}};
    phosphor::dump::createPEL(additionalData,
                              phosphor::dump::PelSeverity::INFORMATIONAL,
                              "xyz.openbmc_project.Dump.Error.Invalidate");
}`
I don't see any sd_bus object in this method to be passed on. Same for system_dump_entry.cpp. Plz correct me if I am wrong.

File: dump_utils.cpp (hunk 5 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:192]: I don't see the caller in resource_dump_entry.cpp and system_dump_entry.cpp has this sd_bus object to be passed on. Please correct me if I am wrong with my analysis.
`void Entry::delete_()
{
    auto dumpId = id;
    // Skip the system dump delete if the dump is in progress
    // and in memory preserving reboot path
    if ((openpower::dump::util::isInMpReboot()) &&
        (status() == phosphor::dump::OperationStatus::InProgress))
    {
        log<level::INFO>(
            fmt::format(
                "Skip deleting system dump delete id({}) durng mp reboot",
                dumpId)
                .c_str());
        return;
    }

    auto srcDumpID = sourceDumpId();

    if ((!offloadUri().empty()) && (phosphor::dump::isHostRunning()))
    {
        log<level::ERR>(
            fmt::format("Dump offload in progress id({}) srcdumpid({})", dumpId,
                        srcDumpID)
                .c_str());
        elog<sdbusplus::xyz::openbmc_project::Common::Error::Unavailable>();
    }

    log<level::INFO>(fmt::format("System dump delete id({}) srcdumpid({})",
                                 dumpId, srcDumpID)
                         .c_str());

    auto path =
        std::filesystem::path(SYSTEM_DUMP_SERIAL_PATH) / std::to_string(dumpId);

    // Remove host system dump when host is up by using source dump id
    // which is present in system dump entry dbus object as a property.
    if ((phosphor::dump::isHostRunning()) && (srcDumpID != INVALID_SOURCE_ID))
    {
        try
        {
            phosphor::dump::host::requestDelete(srcDumpID,
                                                TRANSPORT_DUMP_TYPE_IDENTIFIER);
        }
        catch (const std::exception& e)
        {
            log<level::ERR>(fmt::format("Error deleteing dump from host id({}) "
                                        "host id({}) error({})",
                                        dumpId, srcDumpID, e.what())
                                .c_str());
            elog<sdbusplus::xyz::openbmc_project::Common::Error::Unavailable>();
        }
    }

    // Remove Dump entry D-bus object
    phosphor::dump::Entry::delete_();
    try
    {
        std::filesystem::remove_all(path);
    }
    catch (const std::filesystem::filesystem_error& e)
    {
        // Log Error message and continue
        log<level::ERR>(
            fmt::format("Failed to delete dump file({}), errormsg({})",
                        path.string().c_str(), e.what())
                .c_str());
    }

    log<level::INFO>(
        fmt::format("System dump entry with id({}) is deleted", dumpId)
            .c_str());

    // Log PEL for dump delete/offload
    log<level::INFO>("Log PEL for dump delete or offload");

    std::unordered_map<std::string, std::string> additionalData = {
        {"Dump ID", std::to_string(dumpId)},
        {"Filename", path},
        {"Dump Type", "System Dump"}};
    phosphor::dump::createPEL(additionalData,
                              phosphor::dump::PelSeverity::INFORMATIONAL,
                              "xyz.openbmc_project.Dump.Error.Invalidate");
}`

File: dump_utils.hpp (hunk 4 above)
Comment (Swarnendu-R-C) [reply dump_utils.hpp:193]: Screenshot 2023-01-19 at 5 09 00 PM
Getting compilation error. The sdbusplus wrapper has no direct access to it's internal sd_bus* object

File: dump_utils.cpp (hunk 5 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:192]: Screenshot 2023-01-19 at 5 09 00 PM
Getting compilation error. The sdbusplus wrapper has no direct access to it's internal sd_bus* object and async method is not yet available in the wrapper class. Please correct me if I am wrong with my analysis.

File: dump_utils.cpp (hunk 3 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp]: Can't remove as async method needs a sd_bus* object which is not directly available from bus object. Kindly correct me if I am wrong with my analysis.

File: dump_utils.cpp (hunk 9 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:179]: Hopefully corrected now. Kindly check and confirm. Sorry for the mess up.

File: dump_utils.cpp [hunk 10]
```diff
@@ -147,5 +147,77 @@ bool isHostQuiesced()
...
+        constexpr auto loggerCreateInterface =
+            "xyz.openbmc_project.Logging.Create";
+        constexpr auto loggerService = "xyz.openbmc_project.Logging";
+        std::string pelSeverity =
+            "xyz.openbmc_project.Logging.Entry.Level.Informational";
+        auto itr = sevMap.find(sev);
+        if (itr != sevMap.end())
+            pelSeverity = itr->second;
+
+        sd_bus* pSD_Bus = nullptr;
```
Comment (devenrao) [inline dump_utils.cpp:190]: lowerCamelCase

File: dump_utils.cpp [hunk 11]
```diff
@@ -147,5 +147,62 @@ bool isHostQuiesced()
...
+            pelSevDefault = pelSev;
+        }
+        constexpr auto loggerObjectPath = "/xyz/openbmc_project/logging";
+        constexpr auto loggerCreateInterface =
+            "xyz.openbmc_project.Logging.Create";
+        constexpr auto loggerService = "xyz.openbmc_project.Logging";
+        constexpr auto dumpFileString = "File Name";
+        constexpr auto dumpFileTypeString = "Dump Type";
+        constexpr auto dumpIdString = "Dump ID";
+
```
Comment (devenrao) [inline dump_utils.cpp:179]: That code appears to be completely avoiding using sdbusplus.  No, that is not my preference, and I would not suggest it to be a pattern to be copied.
It's pretty much structured as C code compiled with the C++ compiler.

this is the response from Patrick on discord, so this is not the correct way to do.

Comment (devenrao) [reply dump_utils.cpp:179]: https://github.com/openbmc/sdbusplus/blob/master/example/coroutine-example.cpp

please try out this one

**Comment by geissonator** [comment]
add to approvelist

File: dump_utils.cpp (hunk 10 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp]: Done

File: dump_utils.cpp (hunk 11 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:179]: As per the discussion we are doing the co-routine stuffs in 1050/master at later stage. Going with the current changes as of now

File: dump_utils.hpp [hunk 12]
```diff
@@ -166,5 +166,19 @@ T readDBusProperty(sdbusplus::bus::bus& bus, const std::string& service,
...
 
+/**
+ * @brief Create a new PEL message for dump Delete/Offload
+ *
+ * @param[in] dumpFilePath - Deleted dump file path/name
+ * @param[in] dumpFileType - Deleted dump file type (BMC/Resource/System)
+ * @param[in] dumpId - The dump ID
+ * @param[in] pelSev - PEL severity (Informational by default)
+ * @param[in] errIntf - D-Bus interface name.
+ * @return Returns void on success throws exception on error
```
Comment (devenrao) [inline dump_utils.hpp:177]: this method does not throw any error

File: dump_utils.cpp [hunk 13]
```diff
@@ -147,5 +147,62 @@ bool isHostQuiesced()
...
+    {
+        if (dumpFilePath.empty() || dumpFileType.empty())
+        {
+            log<level::WARNING>(
+                fmt::format("User data is not sufficient to "
+                            "create a PEL message for dump delete/offload")
+                    .c_str());
+            throw std::runtime_error("Dump delete/offload PEL not created due "
+                                     "to insufficient info passed");
+        }
```
Comment (devenrao) [inline dump_utils.cpp:165]: not required, worst case in the pel these fields will be empty, we are not going to crash if they are empty

File: dump_utils.cpp [hunk 14]
```diff
@@ -147,5 +147,62 @@ bool isHostQuiesced()
...
+        std::string pelSevDefault =
+            "xyz.openbmc_project.Logging.Entry.Level.Informational";
+        if (!pelSev.empty())
+        {
+            pelSevDefault = pelSev;
```
Comment (devenrao) [inline dump_utils.cpp:170]: not required

File: dump_utils.hpp (hunk 12 above)
Comment (Swarnendu-R-C) [reply dump_utils.hpp:177]: Changed

File: dump_utils.cpp (hunk 13 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:165]: Agree. Removed

File: dump_utils.cpp (hunk 14 above)
Comment (Swarnendu-R-C) [reply dump_utils.cpp:170]: Removed

**Review by devenrao — APPROVED** [review]
